### PR TITLE
[SPARK-21414] Refine SlidingWindowFunctionFrame to avoid OOM.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

In `SlidingWindowFunctionFrame`, it is now adding all rows to the buffer for which the input row value is equal to or less than the output row upper bound, then drop all rows from the buffer for which the input row value is smaller than the output row lower bound.
This could result in the buffer is very big though the window is small.
For example: 
```
select a, b, sum(a) 
over (partition by b order by a range between 1000000 following and 1000001 following) 
from table
```
We can refine the logic and just add the qualified rows into buffer.

## How was this patch tested?
Manual test:
Run sql 
`select shop, shopInfo, district, sum(revenue) over(partition by district order by revenue range between 100 following and 200 following) from revenueList limit 10` 
against a table with 4  columns(shop: String, shopInfo: String, district: String, revenue: Int). The biggest partition is around 2G bytes, containing 200k lines. 
Configure the executor with 2G bytes memory. 
With the change in this pr, it works find. Without this change, below exception will be thrown.
```
MemoryError: Java heap space
	at org.apache.spark.sql.catalyst.expressions.UnsafeRow.copy(UnsafeRow.java:504)
	at org.apache.spark.sql.catalyst.expressions.UnsafeRow.copy(UnsafeRow.java:62)
	at org.apache.spark.sql.execution.window.SlidingWindowFunctionFrame.write(WindowFunctionFrame.scala:201)
	at org.apache.spark.sql.execution.window.WindowExec$$anonfun$14$$anon$1.next(WindowExec.scala:365)
	at org.apache.spark.sql.execution.window.WindowExec$$anonfun$14$$anon$1.next(WindowExec.scala:289)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:231)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:225)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:341)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
```

### CODE REVIEW

**Comment by jinxing64** [comment]
retest please

**Comment by jinxing64** [comment]
Jenkins, retest this please.

**Comment by jinxing64** [comment]
cc @cloud-fan @jiangxb1987

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 1]
```diff
@@ -212,6 +203,21 @@ private[window] final class SlidingWindowFunctionFrame(
       bufferUpdated = true
     }
 
+    // Add all rows to the buffer for which the input row value is equal to or less than
+    // the output row upper bound.
+    while (nextRow != null && ubound.compare(nextRow, inputHighIndex, current, index) <= 0) {
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala]: How about:
```
     while (nextRow != null && ubound.compare(nextRow, inputHighIndex, current, index) <= 0) {
       if (lbound.compare(nextRow, inputLowIndex, current, index) < 0) {
         inputLowIndex += 1
       } else {
         buffer.add(nextRow.copy())
         bufferUpdated = true
       }
       nextRow = WindowFunctionFrame.getNextOrNull(inputIterator)
       inputHighIndex += 1
     }
```
?

**Review by jiangxb1987 — COMMENTED** [review]
I think this improvement is valid, although this only take effect on relative corner case(when CurrentRow is not in the window frame). One concern is the test case don't reflect the improvement of this change.

**Comment by jinxing64** [comment]
@jiangxb1987 Thanks a lot for quick reply !

> One concern is the test case don't reflect the improvement of this change.

Yes, there is no unit test for `WindowFunctionFrame` now. The test case I added is just to check the correctness.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala [hunk 2]
```diff
@@ -356,6 +356,42 @@ class SQLWindowFunctionSuite extends QueryTest with SharedSQLContext {
     spark.catalog.dropTempView("nums")
   }
 
+  test("window function: mutiple window expressions specified by range in a single expression") {
+    val nums = sparkContext.parallelize(1 to 10).map(x => (x, x % 2)).toDF("x", "y")
+    nums.createOrReplaceTempView("nums")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: wrap your test with `withTempView`, which can drop the view automatically.

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: BTW this test is not very related to this PR, just adds test coverage for range window frame.

**Comment by cloud-fan** [comment]
@jinxing64 I think this patch is straightforward, can you do a manual test, which OOM before and works after this PR? We can put the test in PR description so that other people can try it out.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala (hunk 2 above)
Comment (jiangxb1987) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: And this test case doesn't cover when CurrentRow is not in the window frame. We'd better add that senario.

**Comment by jinxing64** [comment]
@cloud-fan @jiangxb1987 
Thanks for help! I will refine and post the result of manual test late today :)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala (hunk 2 above)
Comment (jinxing64) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: Sure, I will add it later today.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala [hunk 3]
```diff
@@ -359,37 +359,41 @@ class SQLWindowFunctionSuite extends QueryTest with SharedSQLContext {
...
-        |       w2 AS (PARTITION BY y ORDER BY x RANGE BETWEEN 2 PRECEDING AND 2 FOLLOWING),
-        |       w3 AS (PARTITION BY y ORDER BY x RANGE BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING)
-      """.stripMargin
-    )
-
-    checkAnswer(actual, expected)
-    spark.catalog.dropTempView("nums")
+    withTempView("nums") {
+      val expected =
+        Row(1, 1, 1, 4, null, 8, 25) ::
```
Comment (jinxing64) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: @cloud-fan 
This is null and do you think 0 is better?

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala [hunk 4]
```diff
@@ -356,6 +356,46 @@ class SQLWindowFunctionSuite extends QueryTest with SharedSQLContext {
...
+          Row(1, 3, 4, 9, 1, 12, 24) ::
+          Row(1, 5, 9, 15, 4, 16, 21) ::
+          Row(1, 7, 16, 21, 8, 9, 16) ::
+          Row(1, 9, 25, 16, 12, null, 9) ::
+          Row(0, 2, 2, 6, null, 10, 30) ::
+          Row(0, 4, 6, 12, 2, 14, 28) ::
+          Row(0, 6, 12, 18, 6, 18, 24) ::
+          Row(0, 8, 20, 24, 10, 10, 18) ::
+          Row(0, 10, 30, 18, 14, null, 10) ::
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: BTW, please make sure there is no behavior change, i.e. the result should be same with or without this PR.

Comment (jinxing64) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: `expected` is calculated manually. This test is to verify there is no behavior change.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLWindowFunctionSuite.scala]: null is better, which matches the behavior in Aggregate.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!

**Pull request closed without merging**